File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    answer = []
    con = dict()
    for start, end in tickets:
        if start in con:
            con[start].append(end)
        else:
            con[start] = [end]
    
    for c in con.keys():
        con[c].sort(reverse = True)
    stack = ['ICN']
    while len(stack) != 0:
        top = stack[-1]
        if top not in con or not con[top]:
            answer.append(stack.pop())
        else:
            stack.append(con[top].pop())
            print(con)
    return answer[::-1]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    
    static Map<String, PriorityQueue<String>> route = new HashMap<>();
    static List<String> path = new ArrayList<>();
    
    public String[] solution(String[][] tickets) {
        
        for(String[] t : tickets){
            PriorityQueue<String> tmp = route.getOrDefault(t[0], new PriorityQueue<>());
            tmp.offer(t[1]);
            route.put(t[0], tmp);
        }
        
        dfs("ICN");
        Collections.reverse(path);
        return path.toArray(new String[path.size()]);
    }
    
    public void dfs(String from){
        PriorityQueue<String> pq = route.get(from);
        while (pq != null && !pq.isEmpty()) {
            String to = pq.poll();
            dfs(to);
        }
        path.add(from);
    }
}